New hamocc formatting

mvertens opened the pull request:
This PR does the following:
- All hamocc code is now in modules
- There is new consistent indentation across all of the module files
- All implicit none statements are only at the module level
- Save statements at the module level have been removed
- There has been a cleanup of the history documentation blocks at the top of the many of the modules and subroutines

mvertens committed:
rename all files to have module names

mvertens committed:
udpates to convert to modules, indentation and other extensive formatting changes

mvertens committed:
removed files that are now in mo_carchm.F90

mvertens committed:
a few fixes for compilation problems

mvertens commented:
@TomasTorsvik @jmaerz @JorgSchwinger - this PR is a substitution for the previous PR that had problems that @TomasTorsvik brought up. Hopefully, this is a more clean version. Although the PR comes from my fork - there is also a branch in the noresmhub repository that can be looked at.

mvertens committed:
DO, ENDIF, IF, ELSEIF, ENDIF, etc made lower case

jmaerz commented:
@TomasTorsvik and @JorgSchwinger , since we briefly discussed the issue during lunch: it seems as if github isn't correctly showing the renaming. At least when I download Marianas branch and check it with `gitk`, the renaming also properly shows up (similar to when looking into the individual commits on github). Anyway, @mvertens : it is probably good to merge the full commit record (and **not** squash the commits).

jmaerz commented on hamocc/mo_accfields.F90:
```diff
@@ -0,0 +1,463 @@
+  implicit none
+  private
+
+  public :: ACCFIELDS
```
ACCFIELDS could be small letters

mvertens replied:
Thanks. done.

jmaerz commented on hamocc/mo_accfields.F90:
```diff
@@ -0,0 +1,463 @@
+
+    ! Arguments
+    integer , intent(in) :: kpie                  ! 1st dimension of model grid.
+    integer , intent(in) :: kpje                  ! 1st dimension of model grid.
```
comment needs update for kpje and kpke (2. and 3. dimension)

jmaerz commented on hamocc/mo_apply_oafx.F90:
```diff
@@ -3,107 +3,75 @@
+    ! Arguments
+    integer, intent(in) :: kpie                       ! 1st dimension of model grid.
+    integer, intent(in) :: kpje                       ! 2nd dimension of model grid.
+    integer, intent(in) :: kpke                       ! size of grid cell (depth) [m].
```
comments need update: kpke: vertical dimension, pddpo: size of grid cell, omask: loand/ocean mask, oafx: alkalinization field to apply

mvertens replied:
Done. Thanks for catching this.

mvertens commented:
@jmaerz - I'm not sure what you mean by squashing the commit. This PR has implemented the steps that @TomasTorsvik suggested. First I simply renamed the code and committed that. Then I updated the code with the full changes that were part of the previous PR. Then I added lower case changes.

jmaerz commented:
Hi @mvertens , once merging this PR into master, we should make sure that the full commit history is merged and not squashed.

mvertens committed:
addressed changes in the review

mvertens commented:
@jmaerz - thanks for your clarification. That makes total sense.

jmaerz commented on hamocc/mo_aufr_bgc.F90:
```diff
@@ -0,0 +1,581 @@
+    integer   :: restdtoce                         !  time step number from bgc ocean file
+    integer   :: idate(5),i,j,k
+    logical   :: lread_cfc,lread_nat,lread_iso,lread_atm,lread_bro
+    real      :: rco213,rco214,alpha14,beta13,beta14,d13c_atm,d14cat ! cisonew
```
Please refrain from changing variable letters and revert back `d13c_atm` to `d13C_atm` (the capital C has a  clear meaning here). (at least that's my opinion)

mvertens replied:
Sorry about that - this was inadvertent. I'll fix this!

jmaerz replied:
No worries, just commenting and going through your commit. Appreciate your work!

mvertens replied:
Actually - looking at the original code - there are still a lot of places in the original code where d13c is lower case.
Should we change those?

jmaerz replied:
I guess, I leave this to @JorgSchwinger , in my opinion, it could be done, but since the isotopes are not my expertise thus far, I don't want to enforce here something. I just stumbled over it, when looking through your changes.

jmaerz commented on hamocc/mo_aufw_bgc.F90:
```diff
@@ -0,0 +1,931 @@
+       kplyear,kplmon,kplday,kpldtoce,omask,rstfnm)
+
+    !******************************************************************************
+    ! Wwrite marine bgc restart data.
```
Just `Write` (typo)

mvertens replied:
Done.

mvertens committed:
more changes for review

mvertens committed:
more changes for review

jmaerz commented:
Hi @mvertens, @JorgSchwinger , @TomasTorsvik , I recognized that you removed numerous `implicit none` in subroutines. I am a bit uncertain, if we should still maintain the old style (incl. the `implicit none` in each subroutine, if possible, while I understand the intent of yours).

mvertens committed:
more downcase

mvertens commented:
@jmaerz - it is viewed as best practices now to only have implicit none at the module level. 
If you always use implicit none in your Fortran code, you only need to type it in programs and modules. 
If you want me to re-introduce it I will.

mvertens commented:
Just to add that all new code in CESM has adopted this.

mvertens committed:
updated comment

jmaerz commented on hamocc/mo_bgcmean.F90:
```diff
@@ -5,2354 +5,2216 @@
+    ! --- ------------------------------------------------------------------
+    !
+    ! Arguments
+    integer, intent(in) :: pos(nbgcmax)       ! position in 2d buffer
```
`! position in 3d buffer` , right?

mvertens replied:
Yes. Fixed.

jmaerz commented on hamocc/mo_bgcmean.F90:
```diff
@@ -5,2354 +5,2216 @@
+    ! --- Description: writes diagnostic 2d field to file
+    ! --- ------------------------------------------------------------------
+    !
+    ! Arguments
```
Pleasew update the comments according to the former comments:
```
! --- Arguments:
! ---   int  pos      (in)     : variable position in common buffer
! ---   int  frmt     (in)     : format/precision of output 
! ---                            0=field is not written  
! ---                            2=field is written as int2 with scale 
! ---                              factor and offset 
! ---                            4=field is written as real4
! ---                            8=field is written as real8
! ---   real sfac     (in)     : user def.NE. scale factor to be applied   
! ---   real offs     (in)     : user def.NE. offset to be added 
! ---   int  cmpflg   (in)     : compression flag; only wet points are 
! ---                            written IF flag is set to 1 
! ---   char vnm      (in)     : variable name used in nc-file 
```
`sfac`and `pos` mixed up and `offs`, etc.

mvertens replied:
Done.

jmaerz commented on hamocc/mo_bgcmean.F90:
```diff
@@ -5,2354 +5,2216 @@
+    ! --- ------------------------------------------------------------------
+    ! --- Description: writes diagnostic layer field to file
+    ! --- ------------------------------------------------------------------
+    !
```
same here. Former description:
```
! ---   
! --- Arguments:
! ---   int  pos      (in)     : variable position in common buffer
! ---   int  frmt     (in)     : format/precision of output 
! ---                            0=field is not written  
! ---                            2=field is written as int2 with scale 
! ---                              factor and offset 
! ---                            4=field is written as real4
! ---                            8=field is written as real8
! ---   real sfac     (in)     : user def.NE. scale factor to be applied   
! ---   real offs     (in)     : user def.NE. offset to be added 
! ---   int  cmpflg   (in)     : compression flag; only wet points are 
! ---                            written IF flag is set to 1 
! ---   char vnm      (in)     : variable name used in nc-file 
```

mvertens replied:
Done.

jmaerz commented on hamocc/mo_bgcmean.F90:
```diff
@@ -5,2354 +5,2216 @@
+
+
+
+  subroutine wrtlvl(pos,frmt,sfac,offs,cmpflg,vnm)
```
And again.
```
! --- Arguments:
! ---   int  pos      (in)     : variable position in common buffer
! ---   int  frmt     (in)     : format/precision of output 
! ---                            0=field is not written  
! ---                            2=field is written as int2 with scale 
! ---                              factor and offset 
! ---                            4=field is written as real4
! ---                            8=field is written as real8
! ---   real sfac     (in)     : user def.NE. scale factor to be applied   
! ---   real offs     (in)     : user def.NE. offset to be added 
! ---   int  cmpflg   (in)     : compression flag; only wet points are 
! ---                            written IF flag is set to 1 
! ---   char vnm      (in)     : variable name used in nc-file 
```

mvertens replied:
Done.

jmaerz commented on hamocc/mo_bgcmean.F90:
```diff
@@ -5,2354 +5,2216 @@
+
+
+
+  subroutine wrtsdm(pos,frmt,sfac,offs,cmpflg,vnm)
```
And again.
```
! --- Description: writes diagnostic sediment field to file  
! ---   
! --- Arguments:
! ---   int  pos      (in)     : variable position in common buffer
! ---   int  frmt     (in)     : format/precision of output 
! ---                            0=field is not written  
! ---                            2=field is written as int2 with scale 
! ---                              factor and offset 
! ---                            4=field is written as real4
! ---                            8=field is written as real8
! ---   real sfac     (in)     : user def.NE. scale factor to be applied   
! ---   real offs     (in)     : user def.NE. offset to be added 
! ---   int  cmpflg   (in)     : compression flag; only wet points are 
! ---                            written IF flag is set to 1 
! ---   char vnm      (in)     : variable name used in nc-file 
```

mvertens replied:
Done.

jmaerz commented on hamocc/mo_bgcmean.F90:
```diff
@@ -5,2354 +5,2216 @@
+
+
+
+  subroutine wrtbur(pos,frmt,sfac,offs,cmpflg,vnm)
```
And again.
```
! --- Description: writes diagnostic sediment burial field to file  
! ---   
! --- Arguments:
! ---   int  pos      (in)     : variable position in common buffer
! ---   int  frmt     (in)     : format/precision of output 
! ---                            0=field is not written  
! ---                            2=field is written as int2 with scale 
! ---                              factor and offset 
! ---                            4=field is written as real4
! ---                            8=field is written as real8
! ---   real sfac     (in)     : user def.NE. scale factor to be applied   
! ---   real offs     (in)     : user def.NE. offset to be added 
! ---   int  cmpflg   (in)     : compression flag; only wet points are 
! ---                            written IF flag is set to 1 
! ---   char vnm      (in)     : variable name used in nc-file 
```

mvertens replied:
Done.

jmaerz commented on hamocc/mo_bgcmean.F90:
```diff
@@ -5,2354 +5,2216 @@
+    ! --- Description: replace 2d field with log10(field)
+    ! --- ------------------------------------------------------------------
+    !
+    ! Arguments
```
And again.
```
! --- Arguments:
! ---   int  pos      (in)     : field position in layer buffer 
! ---   real sfac     (in)     : scale factor to be applied before log10   
! ---   real offs     (in)     : offset to be added before log10   
```

jmaerz commented on hamocc/mo_bgcmean.F90:
```diff
@@ -5,2354 +5,2216 @@
+    ! --- ------------------------------------------------------------------
+    !
+    ! Arguments
+    real,    intent(in) :: sfac ! field position in layer buffer
```
And again.
```
! --- Arguments:
! ---   int  pos      (in)     : field position in layer buffer 
! ---   real sfac     (in)     : scale factor to be applied before log10   
! ---   real offs     (in)     : offset to be added before log10
```

mvertens replied:
Done.

jmaerz commented on hamocc/mo_bgcmean.F90:
```diff
@@ -5,2354 +5,2216 @@
+    ! --- ------------------------------------------------------------------
+    !
+    ! Arguments
+    real,    intent(in) :: sfac ! field position in layer buffer
```
```
! --- Arguments:
! ---   int  pos      (in)     : field position in layer buffer 
! ---   real sfac     (in)     : scale factor to be applied before log10   
! ---   real offs     (in)     : offset to be added before log10   
```

mvertens replied:
Done.

mvertens replied:
Done.

jmaerz commented on hamocc/mo_bgcmean.F90:
```diff
@@ -5,2354 +5,2216 @@
+    ! --- ------------------------------------------------------------------
+    !
+    ! Arguments
+    real,    intent(in) :: sfac ! field position in layer buffer
```
```
! --- Arguments:
! ---   int  pos      (in)     : field position in layer buffer 
! ---   real sfac     (in)     : scale factor to be applied before log10   
! ---   real offs     (in)     : offset to be added before log10  
```

mvertens replied:
Done.

mvertens replied:
Done.

mvertens commented:
@jmaerz - thank you so much for your careful review. I'll fix all of these tonight.

jmaerz commented on hamocc/mo_bgcmean.F90:
```diff
@@ -5,2354 +5,2216 @@
+                EXIT
+              endif
+              ind2(i,j)=d
+              weights(i,j,d)=(min(zbot(i,j,kin), depthslev_bnds(2,d)) - &
```
This change in line break feels odd - there was a clear meaning before (`depthlev_bnds(2,d) - ` should be placed in the next line)

mvertens replied:
Done. Thanks for pointing this out.

jmaerz commented on hamocc/mo_biomod.F90:
```diff
@@ -4,359 +4,345 @@
+    ! local variables
+    integer :: errstat
+
+    if (mnproc.eq.1) THEN
```
The `THEN` (here and thereafter) could be also written in small letters.

mvertens replied:
I have changed all THEN->then and also addressed your other review comments.

mvertens committed:
clean up of comments and change THEN->then

jmaerz commented on hamocc/mo_carbch.F90:
```diff
@@ -4,419 +4,388 @@
+      write(io_stdo_bgc,*)'First dimension    : ',kpie
+      write(io_stdo_bgc,*)'Second dimension   : ',kpje
+    endif
+    allocate (kwco2sol(kpie,kpje),stat=errstat)
```
Thanks for catching these formerly detached lines.

jmaerz commented on hamocc/mo_carchm.F90:
```diff
@@ -0,0 +1,880 @@
+    real,    intent(in) :: pglat(1-kbnd:kpie+kbnd,1-kbnd:kpje+kbnd)          !  latitude of grid cells [deg north].
+    real,    intent(in) :: omask(kpie,kpje)                                  !  ocean mask.
+    real,    intent(in) :: psicomo(1-kbnd:kpie+kbnd,1-kbnd:kpje+kbnd)        !  sea ice.
+    real,    intent(in) :: ppao(1-kbnd:kpie+kbnd,1-kbnd:kpje+kbnd)           !  sea level presure [pascal].
```
Former typo could be fixed (`pressure` - double `s`)

mvertens replied:
Done.

mvertens committed:
fixed typo in comment

jmaerz commented on hamocc/mo_carchm.F90:
```diff
@@ -0,0 +1,880 @@
+    real    :: ato2,atn2,atco2,pco2
+    real    :: oxy,ani,anisa
+    real    :: rrho,t,t2,t3,t4,tk,tk100,prb,s,rs
+    real    :: kh,khd,k1,k2,kb,k1p,k2p,k3p,ksi,kw,ks1,kf,kspc,kspa
```
Not sure, but also here the capital letters had a meaning (in my opinion)

JorgSchwinger replied:
Yes I agree, generally capital letters in variable names have been there for a good reason and should stay.

mvertens replied:
Right. Fixed.

mvertens committed:
fixed wrong comment

jmaerz commented on hamocc/mo_carchm.F90:
```diff
@@ -0,0 +1,880 @@
+    real,    intent(in)    :: saln   ! salinity [psu].
+    real,    intent(in)    :: prb    ! pressure [bar].
+    real,    intent(out)   :: Kh     ! equilibrium constant Kh  =  [CO2]/pCO2, moist air.
+    real,    intent(out)   :: Khd    ! equilibrium constant Kh  =  [CO2]/pCO2, dry air.
```
comment for `Khd` could be fixed: `Kh` -> `Khd` (formerly wrong)

mvertens replied:
Fixed.

jmaerz replied:
Sorry being nit-picky, but the comment still shows `Kh` instead of `Khd` - anyway - was our fault initially.

mvertens committed:
fixed capitalization problem

mvertens committed:
fixed wrong comment

jmaerz commented on hamocc/mo_carchm.F90:
```diff
@@ -0,0 +1,880 @@
+    real,    intent(in)  :: ta     ! total alkalinity [eq/kg].
+    real,    intent(in)  :: sit    ! silicate concentration [mol/kg].
+    real,    intent(in)  :: pt     ! phosphate concentration [mol/kg].
+    real,    intent(in)  :: KH     ! equilibrium constant K0  = [H2CO3]/pCO2.
```
`KH` -> `Kh` @JorgSchwinger : any reason, why the comment says `K0` and not `Kh`?

mvertens replied:
Fixed for both the comment and the variable in the declaration.

jmaerz commented on hamocc/mo_carchm.F90:
```diff
@@ -0,0 +1,880 @@
+
+  end subroutine carchm_solve_dicsat
+
+end module mo_carchm
```
Hi @TomasTorsvik and @JorgSchwinger , for now, I did not compare the content of the two added subroutines (I considered them as unchanged besides the indentation).

mvertens replied:
They were totally unchanged except for the indentation and other formatting changes that applied to the other routines.

jmaerz commented on hamocc/mo_hamocc4bcm.F90:
```diff
@@ -0,0 +1,397 @@
+    real,    intent(in)    :: pdlxp  (kpie,kpje)                              ! size of grid cell (longitudinal) [m].
+    real,    intent(in)    :: pdlyp  (kpie,kpje)                              ! size of grid cell (latitudinal) [m].
+    real,    intent(in)    :: pddpo  (kpie,kpje,kpke)                         ! size of grid cell (depth) [m].
+    real,    intent(in)    :: prho   (kpie,kpje,kpke)                         ! density [kg/m^3].
```
Dear @mvertens , the unit for density `prho` could be fixed in the comment (formerly wrong): `g/cm^3`

mvertens replied:
Done.

jmaerz commented on hamocc/mo_hamocc4bcm.F90:
```diff
@@ -0,0 +1,397 @@
+    real,    intent(in)    :: dust   (kpie,kpje)                              ! dust deposition flux [kg/m2/month].
+    real,    intent(in)    :: rivin  (kpie,kpje,nriv)                         ! riverine input [kmol m-2 yr-1].
+    real,    intent(in)    :: ndep   (kpie,kpje)                              ! nitrogen deposition [kmol m-2 yr-1].
+    real,    intent(in)    :: oafx   (kpie,kpje)
```
`oafx` comment: `! alkalinity flux from alkalinization [kmol m-2 yr-1]` while `pi_ph` has no comment yet @JorgSchwinger , maybe you can add on this?

mvertens replied:
I think that the comment `! alkalinity flux from alkalinization [kmol m-2 yr-1]`belongs with  the `oafx` variable. Right?

jmaerz replied:
Yes.

jmaerz commented on hamocc/mo_ini_fields.F90:
```diff
@@ -5,324 +5,302 @@
+    ! local variables
+    integer :: i,j,k,l
+    real    :: rco213,rco214,beta13,beta14 ! cisonew
+    real    :: snow ! agg
```
`agg` could be capitalized again, since it refers to the logical `AGG`

mvertens replied:
Done

jmaerz commented on hamocc/mo_param_bgc.F90:
```diff
@@ -100,8 +96,8 @@ module mo_param_bgc
   real, protected :: atm_bromo   = 3.4            ! atmosphere bromophorme concentration
-                                                  ! For now use 3.4ppt from Hense and Quack (2009; Biogeosciences) NEED TO
-                                                  !BE UPDATED WITH Ziska et al. (2013) climatology database
+  ! For now use 3.4ppt from Hense and Quack (2009; Biogeosciences) NEED TO
```
Indentation could be re-gained.

mvertens replied:
Done.

jmaerz commented on hamocc/mo_powach.F90:
```diff
@@ -0,0 +1,519 @@
+    integer, intent(in) :: kpke                                         ! 3rd (vertical) dimension of model grid.
+    integer, intent(in) :: kbnd                                         ! nb of halo grid points
+    real,    intent(in) :: prho(kpie,kpje,kpke)                         ! seawater density [g/cm^3].
+    real,    intent(in) :: omask(kpie,kpje)                             ! salinity [psu].
```
comments for `omask` and `psao` mixed up.

mvertens replied:
Done.

jmaerz commented on hamocc/mo_vgrid.F90:
```diff
@@ -3,334 +3,303 @@
+  real,    dimension(:,:),   allocatable, public :: bolay   ! height of bottom cell.
+  real,    dimension(:,:,:), allocatable, public :: ptiestu ! depth of layer centres.
+  real,    dimension(:,:,:), allocatable, public :: ptiestw ! depth of layer interfaces.
+  integer, dimension(:,:),   allocatable, public :: k0100
```
Could you add the former comment: `k-index of gridbox comprising xxxx m depth`

mvertens replied:
I don't see this comment in the previous code. I have added the following:
  ```
integer, dimension(:,:),   allocatable, public :: k0100   ! k-index of gridbox comprising 100 m depth
  integer, dimension(:,:),   allocatable, public :: k0500   ! k-index of gridbox comprising 500 m depth
  integer, dimension(:,:),   allocatable, public :: k1000   ! k-index of gridbox comprising 1000 m depth
  integer, dimension(:,:),   allocatable, public :: k2000   ! k-index of gridbox comprising 2000 m depth
  integer, dimension(:,:),   allocatable, public :: k4000   ! k-index of gridbox comprising 4000 m depth

```

jmaerz replied:
Perfectly fine as well :-)

jmaerz commented on hamocc/mo_write_netcdf_var.F90:
```diff
@@ -0,0 +1,211 @@
+            stop '(write_netcdf_var)'
+          endif
+
+          ! ncstat=nf90_sync(ncid)
```
Why commented out? - more a question.

mvertens replied:
I'm not sure why this was commented out. @JorgSchwinger @TomasTorsvik - do you know?

jmaerz replied:
Sorry, my bad. I guess, I started to get tired - I overlooked the comment marks at the beginning of the line in the original code. Same applies for below.

jmaerz commented on hamocc/mo_write_netcdf_var.F90:
```diff
@@ -0,0 +1,211 @@
+        stop '(write_pnetcdf_var)'
+      endif
+
+      ! ncstat=nfmpi_sync(ncid)
```
Here again, why commented out?

jmaerz replied:
My bad - I overlooked the original comment signs.

jmaerz commented:
> @jmaerz - it is viewed as best practices now to only have implicit none at the module level. If you always use implicit none in your Fortran code, you only need to type it in programs and modules. If you want me to re-introduce it I will.

Dear @mvertens , I was communicating about this with @TomasTorsvik , the only concern he raised was that when refactoring code, this information could get lost. But granted, with the now happened refactoring, this is probably ruled out. Any further opinion on this @JorgSchwinger @TomasTorsvik ?

jmaerz commented:
Dear @mvertens , so many thanks for your great effort to make the code cleaner and modularize it. I am now once through the code. @TomasTorsvik and @JorgSchwinger , I cannot rule out that some things might have escaped the review, though. There are two things left (from my side). Minor: some files still contain (capital) `THEN`. The other part is: somehow, compiling on github fails, which needs to be solved before being able to merge.

TomasTorsvik commented:
> @jmaerz - it is viewed as best practices now to only have implicit none at the module level. If you always use implicit none in your Fortran code, you only need to type it in programs and modules. If you want me to re-introduce it I will.

I was thinking that it would be good to keep `implicit none` in case we refactor code and move subroutines around. But if we always encapsulate procedures within modules, I agree that we can keep `implicit none` at module level only.

mvertens committed:
changes from review

mvertens commented:
@jmaerz - thanks for your very careful review. I don't see any more instances of THEN. As long as I am still doing this - do you want me to:
replace .GT.  with >
replace .LT. with <
replace .EQ. with ==
replace .NE. with /=

It is trivial to do with across the code base with emacs. So as long as we are changing all the files - I'm happy to add this as well.

jmaerz commented:
According from my trial with `sed` and the comments from @JorgSchwinger  that I received, yes, exchanging `.GT.` etc. would be great.

JorgSchwinger commented:
> > @jmaerz - it is viewed as best practices now to only have implicit none at the module level. If you always use implicit none in your Fortran code, you only need to type it in programs and modules. If you want me to re-introduce it I will.
> 
> I was thinking that it would be good to keep `implicit none` in case we refactor code and move subroutines around. But if we always encapsulate procedures within modules, I agree that we can keep `implicit none` at module level only.

I agree, too.

jmaerz commented:
@mvertens wrt to the compiling issue, could you try to move `subdir('hamocc')` in `./meson.build` (l.156) out of the if for `if get_option('ecosys')` to ca. l. 71 ? - that at least makes the stuff compilable on my local machine again, when `ecosys` is switched off.  - I guess, it's then time to clean `meson.build` a bit... - with the new logicals.

jmaerz commented on hamocc/mo_bgcmean.F90:
```diff
@@ -5,2354 +5,2217 @@
+    !
+    ! Arguments
+    integer, intent(in) :: pos          ! variable position in common buffer
+    real,    intent(in) :: frmt         ! format/precision of output
```
Dear @mvertens , here and below: it seems as if the the type of `frmt` and `offs` got mixed up. (requires: `real :: offs`, `integer :: frmt` )

mvertens replied:
Sorry - I've been recompiling frequently - but missed doing it for this. Should be fixed now.

jmaerz replied:
Could you push the changes?

jmaerz commented:
Me again, sorry. The whole `ecosys` in `./meson.build` can likely be reduced to:
```
subdir('hamocc')

if get_option('ecosys')
  add_project_arguments('-DHAMOCC', language: 'fortran')
endif
```
to switch on the one remained pre-processor flag (in BLOM to switch on HAMOCC).

mvertens committed:
more formatting updates - LT, GT, EQ, NE, LE, GE, etc to lower case and update to meson.build

mvertens commented:
@jmaerz - I have made your suggested change to meson.build and pushed it back along with a lot of the other formatting changes.

jmaerz commented:
getting an:
```
../hamocc/mo_write_netcdf_var.F90:209:5:

  209 |   end subroutine write_netcdf_var
      |     1
Error: Expecting END IF statement at (1)
../hamocc/mo_write_netcdf_var.F90:211:3:

  211 | end module mo_write_netcdf_var
      |   1
Error: Expecting END IF statement at (1)
f951: Error: Unexpected end of file in ‘../hamocc/mo_write_netcdf_var.F90’
```
on my side.

mvertens committed:
fixed placement of ifdef

mvertens commented:
@jmaerz - I pushed a fix back.

mvertens commented:
The reason I did not catch this is because I'm compiling with  -D_PNETCDF and you were not. This should work for you now.

jmaerz commented:
Yeah, teamwork - running the stuff locally at the moment. Get some more issues... See also the github errors.

jmaerz commented:
Line length issues again. The `&` are above the 132 limit in `mo_ocprod` - checking for more - once `mo_ocprod` is fixed, it runs on my side.

mvertens commented:
Just fixed mo_ocprod.F90

jmaerz commented:
Can you push it? - this should be all it needs to get the build system through.

mvertens commented:
I've fixe two more files that are on the edge - and I'm verifying that it compiles right now. I'll push in a minute.

mvertens committed:
fixed line lengths that were very long

jmaerz commented:
:-D - sorry, just curious at the moment ... - didn't mean to be pushy.

mvertens commented:
@jmaerz - not push at all! I really appreciate your help!

jmaerz commented:
Ok, runs through locally now :+1:

mvertens commented:
Great. But github actions is still showing compilation errors.

mvertens commented:
Maybe I spoke too soon?

jmaerz commented:
Yes, I see this happen - it is likely related to the transfer of model parameters into mo_param_bgc and the namelist. For now, we could leave the parameters out of the namelist in `mo_param_bgc` - it is related to #294

mvertens commented:
Should that be part of this PR?

jmaerz commented:
It's a bit of a tricky task, since these values are also used and changed locally (and used in OMP directives) and we thought about fixing this in the `beyond-CMIP6` branch after merging master... - but you're right, maybe we should do a separate attempt to fix this before merging.

mvertens commented:
I'm happy to help with this. Can we talk about this now or in the morning? Whatever works better for you.

jmaerz commented:
I feel it would be good to communicate with @JorgSchwinger and potentially @TomasTorsvik about it. It would mean to rename `wpoc`, `wopal`, `wcalc`, `wdust` to `wxxxx_fix` or `wxxx_const` and then add local variables in `mo_ocprod` that have the names `wpoc`, `wopal`, etc. - then rewriting the `wpoc`, etc. for the fix values in the sinking routine. I was previously hesitant to do this, since it touches the OMP directives with which I had formerly issues and slight answer changes - that whole part experiences fixes in `beyond-CMIP6` anyway (since the OMP are not carefully written + there is a bug in the sinking routine - answer changing for the dust - and thus potentially/likely sediment).

So for now, the easier way might be to remove these variables from the namelist read in `mo_param_bgc`.

mvertens commented:
@jmaerz - let me make sure I understand what you are saying. 
- wcal in mo_ocprod.F90 is obtained via a use statement from mo_param_bgc.F90 (based on namelist input) but then is 
```
changed in statements like:
         if (use_AGG) then
                ..... 
                wcal   = wmass(i,j,k) <====
                ......
```

This is really confusing to me. Why would you change a namelist value you read in?
So do we want to make wcal a local variable variable in mo_ocprod.F90 - say call it wcal_loc and use it there?
And then have logic like the following

```
if (.not. use_AGG) then
    wcal_loc = wcal
endif
if (use_AGG) then
   .....
   wcal_loc = wmass(i,j,k)
   ...
else if (use_WLIN) then
   wcald = wcal_loc
else
   wcald = wcal_loc
endif
```

The private variable would be wcal_loc not wcal.
Do you agree with this?

jmaerz commented:
I sent you a zoom invitation

jmaerz commented:
@TomasTorsvik and @JorgSchwinger , Mariana and I, we agreed on taking the sinking routine issue out to another PR - she'll set up a branch (we discussed on how to do it) - and then see, whether it breaks CMIP6 backwards compatibility or not. If not, it can go in to current `master`, otherwise, it will be a fix for the the `beyond-CMIP6` branch as already laid out in #294.

jmaerz reviewed (approved):
Dear Mariana, many thanks again for this great effort to get to a nicer and cleaner code base for HAMOCC. After all the exchange, I am happy to approve the PR.

TomasTorsvik commented:
Thanks for catching the issue with `wcal` and friends.
I agree that this is best addressed in a separate bug fix PR.

jmaerz commented:
Dear @mvertens , I couldn't hold back from having a short other look into the routines this morning since I was for reviewing checking out a particular code version which you then changed afterward. Anyway, I just did a:
```
 grep -E '\b[[:upper:]]+\b' *F90
```
(not the most selective grep, I know, but provides an overview...) in the `./hamocc/` folder and there are a number of instances, where fortran key words still appear in capital fonts (END IF, MAX, SQRT, DATA, DIMENSION, ...). We can change this afterward, but if you want to have another look, go ahead.

jmaerz commented on hamocc/mo_accfields.F90:
```diff
@@ -0,0 +1,495 @@
+    ! Arguments
+    integer , intent(in) :: kpie                  ! 1st dimension of model grid.
+    integer , intent(in) :: kpje                  ! 2nd dimension of model grid.
+    integer , intent(in) :: kpke                  ! size of grid cell (3rd dimension) [m].
```
`kpke` is the 3rd (vertical) dimension of the model grid (no units).

mvertens replied:
Thanks so much. I'll fix these this morning. I'm working on the bug fix as well.

mvertens replied:
Done.

jmaerz commented:
Hi @mvertens , following up on your `mo_ocprod` fix comment in #306 , I would still suggest to do a separate PR for the `mo_ocprod` fix - while, since it seems to be bfb, it can enter `master`.

mvertens committed:
fixed comment

mvertens commented:
I think all of the comments have been addressed for this PR.

jmaerz commented:
> Dear @mvertens , I couldn't hold back from having a short other look into the routines this morning since I was for reviewing checking out a particular code version which you then changed afterward. Anyway, I just did a:
> 
> ```
>  grep -E '\b[[:upper:]]+\b' *F90
> ```
> 
> (not the most selective grep, I know, but provides an overview...) in the `./hamocc/` folder and there are a number of instances, where fortran key words still appear in capital fonts (END IF, MAX, SQRT, DATA, DIMENSION, ...). We can change this afterward, but if you want to have another look, go ahead.

@mvertens Do you want to leave those for later? - just asking :-) - I am totally fine doing this later.

mvertens committed:
more downcasing caught in the review

mvertens commented:
@jmaerz - I've just pushed back the downcasing for MAX, MIN, DATA, SQRT and DIMENSION. Sorry for forgetting about those.

JorgSchwinger reviewed (approved):
Dear @mvertens I have not looked through this PR in detail now, but I plan to have a more detailed look once this is merged and fix remaining issues (if any) with indentation etc. So I approve this for now. Thanks a lot for all this great work!

jmaerz commented:
Dear @mvertens many thanks again! - I just want to confirm that the code also runs still locally and I would say, the changes can now be merged. Just as a gentle reminder - don't squash it to keep the renaming history. Thereafter #307 can go in (maybe it needs another update with `master` before - we'll see).
